### ISSUE TITLE
Cross-platform cross-compilation consultation

### ISSUE THREAD

**Issue by javacode123** [prisma/prisma-engines#4431]
Hello everyone, I am a novice Rust developer. I am currently looking to fork a copy of schem-engine and add some features. I encountered some problems when using GitHub Action for cross-compilation and cannot compile the arm64 x (Windows, Linux, Darwin) product. I would like to ask how the current schem-engine is cross-compiled for 'x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-pc-windows-msvc, aarch64-pc-windows-msvc, x86_64-apple-darwin, aarch64-apple-darwin', and if I could refer to it.

**Comment by janpio** [comment]
This is a screenshot of our internal build CI, which shows all the build we do and publish:
<img width="879" alt="image" src="https://github.com/prisma/prisma-engines/assets/183673/d2681147-93bd-45ed-921b-7572de114420">

Which ones are you interested in?

**Comment by javacode123** [comment]
> This is a screenshot of our internal build CI, which shows all the build we do and publish: <img alt="image" width="879" src="https://user-images.githubusercontent.com/183673/282034117-d2681147-93bd-45ed-921b-7572de114420.png">
> 
> Which ones are you interested in?

@janpio Cool!. Could you please give me the whole scrip, i will chose which i need. thx.

**Comment by janpio** [comment]
No, this is not a script but a whole CLI. If you tell me which ones you are interested in, I can figure out how they are built and share that with you.

**Comment by javacode123** [comment]
> No, this is not a script but a whole CLI. If you tell me which ones you are interested in, I can figure out how they are built and share that with you.

@janpio OK. I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 . Currently i have compile 「aarch64-apple-darwin，x86_64-apple-darwin，x86_64-unknown-linux-gnu，x86_64-pc-windows-msvc 」by githhub flow action by 「windows-latest，ubuntu-latest，macos-latest」. But when i cross compile aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」  on 「window, ubuntn」it shows
```
on ubuntu, cargo build --target aarch64-unknow-linux-gnu,  failed to run custom build command for `openssl-sys v0.9.91`
on windows, cargo build --target aarch64-pc-windows-msvc,  failed to run custom build command for `ring v0.16.20`
```

**Comment by javacode123** [comment]
> > No, this is not a script but a whole CLI. If you tell me which ones you are interested in, I can figure out how they are built and share that with you.
> 
> @janpio OK. I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 . Currently i have compile 「aarch64-apple-darwin，x86_64-apple-darwin，x86_64-unknown-linux-gnu，x86_64-pc-windows-msvc 」by githhub flow action by 「windows-latest，ubuntu-latest，macos-latest」. But when i cross compile aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 on 「window, ubuntn」it shows
> 
> ```
> on ubuntu, cargo build --target aarch64-unknow-linux-gnu,  failed to run custom build command for `openssl-sys v0.9.91`
> on windows, cargo build --target aarch64-pc-windows-msvc,  failed to run custom build command for `ring v0.16.20`
> ```

I also want to know whether I can use this CLI to compile my Rust code. If yes, how can I use it?

**Comment by janpio** [comment]
> I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 .

What names from the screenshot do these refer to?
Our Windows build is here: https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml

We generally build for Linux via Docker with custom images. I pulled all our build commands into this public repository and run them on GH Actions: https://github.com/prisma/prisma-engines-builds/pull/1 These GH jobs all have a "Show build command" step that you can use to copy/paste the command that should just work on a Linux machine.

> I also want to know whether I can use this CLI to compile my Rust code. If yes, how can I use it?

What do you refer to by "this CLI"?

**Comment by javacode123** [comment]
> > I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 .
> 
> What names from the screenshot do these refer to? Our Windows build is here: https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml
> 
> We generally build for Linux via Docker with custom images. I pulled all our build commands into this public repository and run them on GH Actions: [prisma/prisma-engines-builds#1](https://github.com/prisma/prisma-engines-builds/pull/1) These GH jobs all have a "Show build command" step that you can use to copy/paste the command that should just work on a Linux machine.
> 
> > I also want to know whether I can use this CLI to compile my Rust code. If yes, how can I use it?
> 
> What do you refer to by "this CLI"?

@janpio Cool! I think i can compile to all platform by [linux action](https://github.com/prisma/prisma-engines-builds/pull/1) && [windows action](
) && [drawin action](https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-apple-intel.yml;). It seems that for windows, just support x86_64 ?

**Comment by janpio** [comment]
Yes.

**Comment by javacode123** [comment]
> Did this all work for you @javacode123?

Yes. Thx.

**Comment by javacode123** [comment]
> > > I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 .
> > 
> > 
> > What names from the screenshot do these refer to? Our Windows build is here: https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml
> > We generally build for Linux via Docker with custom images. I pulled all our build commands into this public repository and run them on GH Actions: [prisma/prisma-engines-builds#1](https://github.com/prisma/prisma-engines-builds/pull/1) These GH jobs all have a "Show build command" step that you can use to copy/paste the command that should just work on a Linux machine.
> > > I also want to know whether I can use this CLI to compile my Rust code. If yes, how can I use it?
> > 
> > 
> > What do you refer to by "this CLI"?
> 
> @janpio Cool! I think i can compile to all platform by [linux action](https://github.com/prisma/prisma-engines-builds/pull/1) && [windows action](https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml) && [drawin action](https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-apple-intel.yml;). It seems that for windows, just support x86_64 ?

hi, i want to know {/l**inux-musl**/query-engine.gz} use which docker image @janpio

**Comment by javacode123** [comment]
> > > > I want to know how to compile project with target 「aarch64-pc-windows-msvc，aarch64-unknown-linux-gnu 」 .
> > > 
> > > 
> > > What names from the screenshot do these refer to? Our Windows build is here: https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml
> > > We generally build for Linux via Docker with custom images. I pulled all our build commands into this public repository and run them on GH Actions: [prisma/prisma-engines-builds#1](https://github.com/prisma/prisma-engines-builds/pull/1) These GH jobs all have a "Show build command" step that you can use to copy/paste the command that should just work on a Linux machine.
> > > > I also want to know whether I can use this CLI to compile my Rust code. If yes, how can I use it?
> > > 
> > > 
> > > What do you refer to by "this CLI"?
> > 
> > 
> > @janpio Cool! I think i can compile to all platform by [linux action](https://github.com/prisma/prisma-engines-builds/pull/1) && [windows action](https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-windows.yml) && [drawin action](https://github.com/prisma/prisma-engines/blob/main/.github/workflows/build-apple-intel.yml;). It seems that for windows, just support x86_64 ?
> 
> hi, i want to know {/l**inux-musl**/query-engine.gz} use which docker image @janpio

mock get bin data from this https://github.com/steebchen/prisma-client-go/issues/1107

**Comment by aqrln** [comment]
> hi, i want to know {/l**inux-musl**/query-engine.gz} use which docker image

See https://github.com/prisma/engine-images/tree/main/musl for x86_64 and https://github.com/prisma/engine-images/blob/main/cross/linux-musl-arm.Dockerfile for aarch64
